In [2]:
import pandas as pd
import numpy as np

In [13]:
df = pd.read_csv('D:/credit_risk_ml/data/raw/accepted_2007_to_2018Q4.csv')

C:\Users\sagar\AppData\Local\Temp\ipykernel_1944\3945964872.py:1: DtypeWarning: Columns (0,19,49,59,118,129,130,131,134,135,136,139,145,146,147) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('D:/credit_risk_ml/data/raw/accepted_2007_to_2018Q4.csv')


In [14]:
df.shape

(2260701, 151)

<bound method NDFrame.describe of                                                        id  member_id  \
0                                                68407277        NaN   
1                                                68355089        NaN   
2                                                68341763        NaN   
3                                                66310712        NaN   
4                                                68476807        NaN   
...                                                   ...        ...   
2260696                                          88985880        NaN   
2260697                                          88224441        NaN   
2260698                                          88215728        NaN   
2260699  Total amount funded in policy code 1: 1465324575        NaN   
2260700   Total amount funded in policy code 2: 521953170        NaN   

         loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  \
0           3600.0       36

In [16]:
# get the count of each data type

type_counts = df.dtypes.value_counts()
print("--- Data Type Counts ---")
print(type_counts)

--- Data Type Counts ---
float64    113
object      38
Name: count, dtype: int64


In [17]:
df.sample(10)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
2192994   94128377        NaN     4800.0       4800.0           4750.0   
1545495  133106908        NaN     7000.0       7000.0           7000.0   
1198770   25186561        NaN     8000.0       8000.0           8000.0   
1327881   11675631        NaN    18000.0      18000.0          17975.0   
1361818  144649580        NaN     7000.0       7000.0           7000.0   
367298    41162347        NaN    12000.0      12000.0          12000.0   
476348   127642553        NaN    15000.0      15000.0          15000.0   
833711   137681628        NaN    10000.0      10000.0          10000.0   
2107953  122653942        NaN    12800.0      12800.0          12800.0   
1851215    4765117        NaN    24000.0      24000.0          24000.0   

               term  int_rate  installment grade sub_grade  ...  \
2192994   36 months     16.99       171.11     D        D1  ...   
1545495   36 months     14.03       239.35     C        C2  ...   
1198770   36 months      6.03       243.49     A        A1  ...   
1327881   36 months      9.67       578.03     B        B1  ...   
1361818   36 months     10.33       226.96     B        B1  ...   
367298    60 months     13.99       279.16     C        C4  ...   
476348    60 months      9.93       318.19     B        B2  ...   
833711    36 months      8.46       315.50     A        A5  ...   
2107953   60 months     25.82       381.88     E        E4  ...   
1851215   36 months     13.11       809.93     B        B4  ...   

        hardship_payoff_balance_amount hardship_last_payment_amount  \
2192994                            NaN                          NaN   
1545495                            NaN                          NaN   
1198770                            NaN                          NaN   
1327881                            NaN                          NaN   
1361818                            NaN                          NaN   
367298                             NaN                          NaN   
476348                             NaN                          NaN   
833711                             NaN                          NaN   
2107953                            NaN                          NaN   
1851215                            NaN                          NaN   

        disbursement_method  debt_settlement_flag debt_settlement_flag_date  \
2192994                Cash                     N                       NaN   
1545495                Cash                     N                       NaN   
1198770                Cash                     N                       NaN   
1327881                Cash                     N                       NaN   
1361818                Cash                     N                       NaN   
367298                 Cash                     N                       NaN   
476348                 Cash                     N                       NaN   
833711                 Cash                     N                       NaN   
2107953                Cash                     N                       NaN   
1851215                Cash                     N                       NaN   

        settlement_status settlement_date settlement_amount  \
2192994               NaN             NaN               NaN   
1545495               NaN             NaN               NaN   
1198770               NaN             NaN               NaN   
1327881               NaN             NaN               NaN   
1361818               NaN             NaN               NaN   
367298                NaN             NaN               NaN   
476348                NaN             NaN               NaN   
833711                NaN             NaN               NaN   
2107953               NaN             NaN               NaN   
1851215               NaN             NaN               NaN   

        settlement_percentage settlement_term  
2192994                   NaN             NaN  
1545495              

In [18]:
df.tail(10)

id  member_id  \
2260691                                          89996426        NaN   
2260692                                          90006534        NaN   
2260693                                          89955820        NaN   
2260694                                          89885898        NaN   
2260695                                          88977788        NaN   
2260696                                          88985880        NaN   
2260697                                          88224441        NaN   
2260698                                          88215728        NaN   
2260699  Total amount funded in policy code 1: 1465324575        NaN   
2260700   Total amount funded in policy code 2: 521953170        NaN   

         loan_amnt  funded_amnt  funded_amnt_inv        term  int_rate  \
2260691    32000.0      32000.0          32000.0   60 months     14.49   
2260692    16000.0      16000.0          16000.0   60 months     12.79   
2260693    24000.0      24000.0          24000.0   60 months     10.49   
2260694    24000.0      24000.0          24000.0   60 months     12.79   
2260695    24000.0      24000.0          24000.0   60 months     10.49   
2260696    40000.0      40000.0          40000.0   60 months     10.49   
2260697    24000.0      24000.0          24000.0   60 months     14.49   
2260698    14000.0      14000.0          14000.0   60 months     14.49   
2260699        NaN          NaN              NaN         NaN       NaN   
2260700        NaN          NaN              NaN         NaN       NaN   

         installment grade sub_grade  ... hardship_payoff_balance_amount  \
2260691       752.74     C        C4  ...                            NaN   
2260692       362.34     C        C1  ...                            NaN   
2260693       515.74     B        B3  ...                            NaN   
2260694       543.50     C        C1  ...                            NaN   
2260695       515.74     B        B3  ...                            NaN   
2260696       859.56     B        B3  ...                            NaN   
2260697       564.56     C        C4  ...                            NaN   
2260698       329.33     C        C4  ...                            NaN   
2260699          NaN   NaN       NaN  ...                            NaN   
2260700          NaN   NaN       NaN  ...                            NaN   

        hardship_last_payment_amount disbursement_method  \
2260691                          NaN                Cash   
2260692                          NaN                Cash   
2260693                          NaN                Cash   
2260694                          NaN                Cash   
2260695                          NaN                Cash   
2260696                          NaN                Cash   
2260697                          NaN                Cash   
2260698                          NaN                Cash   
2260699                          NaN                 NaN   
2260700                          NaN                 NaN   

         debt_settlement_flag debt_settlement_flag_date settlement_status  \
2260691                     N                       NaN               NaN   
2260692                     N                       NaN               NaN   
2260693                     N                       NaN               NaN   
2260694                     N                       NaN               NaN   
2260695                     N                       NaN               NaN   
2260696                     N                       NaN               NaN   
2260697                     Y                  Mar-2019            ACTIVE   
2260698                     N                       NaN               NaN   
2260699                   NaN                       NaN               NaN   
2260700                   NaN                       NaN               NaN   

        settlement_date settlement_amount settlement_percentage  \
2260691             NaN           

In [21]:
#this wont work because the number of cols are extermly large
df.isnull().sum()

id                             0
member_id                2260701
loan_amnt                     33
funded_amnt                   33
funded_amnt_inv               33
                          ...   
settlement_status        2226455
settlement_date          2226455
settlement_amount        2226455
settlement_percentage    2226455
settlement_term          2226455
Length: 151, dtype: int64

In [22]:
# cols missing valuse more than 40%

missing_pct = df.isnull().mean()*100

# filter
high_missing_cols = missing_pct[missing_pct > 40].sort_values(ascending=False)

# Display the results
print(f"Total columns with >30% missing data: {len(high_missing_cols)}")
print("-" * 30)
print(high_missing_cols)

Total columns with >30% missing data: 46
------------------------------
member_id                                     100.000000
orig_projected_additional_accrued_interest     99.617331
payment_plan_start_date                        99.517097
hardship_length                                99.517097
hardship_dpd                                   99.517097
hardship_amount                                99.517097
deferral_term                                  99.517097
hardship_last_payment_amount                   99.517097
hardship_end_date                              99.517097
hardship_start_date                            99.517097
hardship_reason                                99.517097
hardship_status                                99.517097
hardship_type                                  99.517097
hardship_loan_status                           99.517097
hardship_payoff_balance_amount                 99.517097
settlement_date                                98.485160
settlement_perce

In [23]:
# Identify columns with low missingness (the reliable ones)
reliable_cols = missing_pct[missing_pct <= 40].index.tolist()

print(f"Total Reliable Columns: {len(reliable_cols)}")
print("---")
print(reliable_cols)

Total Reliable Columns: 105
---
['id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'fico_range_low', 'fico_range_high', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'collections_12_mths_ex_med', 'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m', 'total_bal_il', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc'

In [25]:
print(len(reliable_cols))

105


Conclusion

This dataset consists of 2,260,701 loan records with 151 columns, making it computationally heavy and unsuitable for naive, full-memory processing. As a result, careful column triage and feature planning are essential before any preprocessing or modeling.

An initial missing-value audit revealed that 46 columns contain more than 40% missing data, with several exceeding 90%. These columns are largely associated with hardship programs, settlement information, secondary applicants, and free-text descriptions, indicating that their absence is mostly structural or conditional rather than random. Such columns are either applicable to a small subset of borrowers or represent post-loan events, making them poor candidates for general modeling.

Among the remaining 105 relatively reliable columns, features were classified based on their functional role rather than data type. This analysis identified:

Identifier and internal tracking variables (e.g., id, member_id, url, policy_code) that do not carry predictive value.

A clearly defined target variable (loan_status), which will be isolated from all feature engineering steps.

Multiple data leakage–prone variables (e.g., payment totals, recoveries, outstanding principal) that reflect post-loan outcomes and must be excluded to preserve model validity.

A strong set of core numeric financial features representing borrower income, credit history, and loan characteristics.

Several categorical application features suitable for later encoding.

A subset of high-cardinality textual or location-based columns (e.g., job titles, descriptions, ZIP codes) that require special handling or removal due to sparsity and encoding cost.

Date-based variables were also identified and reserved for future feature extraction rather than direct use.

Overall, this notebook establishes a clear and defensible feature selection strategy, reduces modeling risk due to leakage, and prepares the dataset for efficient downstream processing. The outcome of this analysis is a curated feature universe, which will be formally reduced and cleaned in the next stage (02_data_cleaning.ipynb). This structured approach ensures scalability, interpretability, and alignment with real-world credit risk modeling practices.

identifiers & Non-Feature Columns (DROP LATER)

1. id
2. member_id (already 100% missing)
3. url
4. policy_code

📌 Decision:

These columns are identifiers or internal tracking variables and will be excluded from modeling.

------------Pending--------------------

Target Variable

Target: loan_status
1. Expected Loss (EL)
2. Interest rate spread
3. Credit score proxy

loan_status is the target variable and will not be used as a feature.

Leakage-Prone Columns (VERY IMPORTANT)

These columns leak future information.
Using them would be cheating.

From reliable list, these are dangerous:

1. out_prncp

2. out_prncp_inv

3. total_pymnt

4. total_pymnt_inv

5. total_rec_prncp

6. total_rec_int

7. total_rec_late_fee

8. recoveries

9. collection_recovery_fee

10. last_pymnt_amnt

11. last_pymnt_d

12. next_pymnt_d

📌 Decision:

These variables are post-loan outcome variables and will be excluded to prevent data leakage.